[FIX] web_editor: fix bugs in the new banner command

dhba-odoo opened the pull request:
**Current behavior before PR:**

- Inserting a banner after already existing content it's impossible to
  go to the next line of the banner.
- Inserting a banner did not set focus inside the banner; instead, focus
  was placed after the banner.
- Performing Ctrl+A and backspace inside a banner would unintentionally remove
  Its first element, which could be a block tag
- When a banner was inserted at the first line and Ctrl+A and backspace
  is performed, nothing happens.

**Desired behavior after PR is merged:**

- It's now possible to navigate to the next line of the banner.
- Now, upon creating a banner, focus is now correctly set inside the
  banner.
- Now performing Ctrl+A and backspace inside the banner will no longer remove its
  first element if it's a block tag.
-  Now, all the content except banner will be selected and removed.

task-3432167

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/129874).

shsa-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js:
```diff
@@ -100,7 +99,7 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false,
             parentEl.parentElement &&
             !parentEl.parentElement.isContentEditable &&
-            paragraphRelatedElements.includes(this.tagName) &&
+            this.tagName === 'P' &&
```
only 'P' not other block tags?

dhba-odoo replied:
@shsa-odoo  No, only the 'P' tag. otherwise, if the first element inside the banner is a header or any block tag then on backspace it will not convert that tag into a P tag.

shsa-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2283,13 +2288,15 @@ export class Wysiwyg extends Component {
                     <div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-${alertClass} pb-0 pt-3" role="status" data-oe-protected="true">
                         <i class="fs-4 fa ${iconClass} mb-3" aria-label="${_t(title)}"></i>
-                        <div class="o_editable o_editable_no_shadow w-100 ms-3" data-oe-protected="false">
+                        <div class="o_editable_no_shadow w-100 ms-3" data-oe-protected="false" contenteditable="true">
```
purpose?

shsa-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2283,13 +2288,15 @@ export class Wysiwyg extends Component {
                     </div>
                 `).childNodes[0];
+
                 this.odooEditor.execCommand('insert', bannerElement);
-                setSelection(bannerElement.querySelector('.o_editor_banner > div'), 0);
+                setSelection(bannerElement.querySelector('.o_editor_banner > div > p'), 0);
+
```
unnecessary diff.

dhba-odoo replied:
@shsa-odoo Changes Done!

shsa-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -4335,6 +4335,10 @@ export class OdooEditor extends EventTarget {
         ) {
             this.setContenteditableLink(link);
         }
+        const oEditableBanner = closestElement(ev.target, '.o_editable_no_shadow');
+        if (oEditableBanner && !oEditableBanner.isContentEditable) {
+            oEditableBanner.setAttribute('contenteditable', 'true');
+        }
```
IMO, naming it `banner` would be good. Also does our banner will be contenteditable= false all the time? if yes then i thisnk we dont need to check  for this `!oEditableBanner.isContentEditable`. and this part of code will simply look like:
`       const banner = closestElement(ev.target, '.o_editable_no_shadow');
        if (banner) {
            banner.setAttribute('contenteditable', 'true');
        }
`

dhba-odoo replied:
@shsa-odoo When an already existing record is opened or when we refresh the record, only during that time `contentEditable` will be `false` otherwise, it will always be `true`.

shsa-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:
```diff
@@ -1843,6 +1843,11 @@ export function splitElement(element, offset) {
         index < offset ? before.appendChild(child) : after.appendChild(child);
         index++;
     }
+    const blockEl = closestBlock(after);
+    if (isShrunkBlock(blockEl)) {
+        const br = document.createElement('br');
+        after.appendChild(br);
+    }
```
Give some scenario why are we adding this, in form of docString.

dhba-odoo replied:
@shsa-odoo Done!

Zynton commented:
@dhba-odoo Before I go into the details of your PR, could you please split your changes into (ideally) one commit per fix?

With everything mixed together:
- it's harder to keep track of what does what
- in the future if we blame a line to see what it does it will be harder to know what it was for
- if we need to revert a change later, we won't be able to just target one commit

Then you can also add tests in each commit for the specific bug it tackles.
Thank you.

Zynton commented:
Thanks @dhba-odoo. Runbot is red though.

dhba-odoo commented:
@Zynton I've rebased the PR, and Runbot is now green. Thanks!

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2197,13 +2197,13 @@ export class Wysiwyg extends Component {
                 const bannerElement = parseHTML(`
```
About the commit message. Please rename "Heading 1 placeholder" into "Heading 1 Powerbox hint".

Zynton replied:
Is this about a commit that disappeared?

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2197,13 +2197,13 @@ export class Wysiwyg extends Component {
                     <div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-${alertClass} pb-0 pt-3" role="status" data-oe-protected="true">
                         <i class="fs-4 fa ${iconClass} mb-3" aria-label="${_t(title)}"></i>
-                        <div class="w-100 ms-3" data-oe-protected="false">
+                        <div class="w-100 ms-3" data-oe-protected="false" contenteditable="true">
```
Why is this needed? I don't see anything making the banner not editable in the first place, and we sure don't want this to remain editable when the editor is in readonly mode.
If it's just to fix the Powerbox hint, it's the wrong way to fix it. Is it possible to edit the paragraph before your fix?

dhba-odoo replied:
@Zynton I have implemented changes to ensure that when a banner is created, the cursor will now appear inside the banner. Previously, before the merge of this [PR](https://github.com/odoo/odoo/pull/143362/files) the cursor automatically appeared inside the banner upon its creation. However, this functionality was affected after the removal of 'o_editable.'

Regarding the issue related to the powerbox hint, the problem arose because the selection was set on the 'div' element instead of the 'p' tag.

If we dont want the selection inside the banner when it is created, then their is no need for this changes. Additionally, the line of code 'setSelection(bannerElement.querySelector('.o_editor_banner > div'), 0);' is no longer necessary, as it was responsible for setting the selection inside the banner. Therefore, this line can be removed

Zynton replied:
Would it work if you kept your `setSelection` change and removed the `contenteditable="true"` ? I would expect it to if the banner is in the root editable. If not, how come?

dhba-odoo replied:
`<div class="o_editor_banner o_not_editable lh-1 d-flex align-items-center alert alert-${alertClass} pb-0 pt-3" role="status" data-oe-protected="true">` is `contentEditable=False` and because of that `setSelection` will not set selection within the `p` tag.
It seems that `o_not_editable` class is responsible for setting `contentEditable` to false.

Zynton replied:
Ah thank you I hadn't noticed the `.o_not_editable` on the banner!
The [PR you linked to](https://github.com/odoo/odoo/pull/143362) explicitly adds the `div` to which you applied `contenteditable="true"` to the array returned by `getContentEditableAreas`. Then, `_activateContenteditable` is supposed to apply `contenteditable="true"` to it automatically (and `_stopContenteditable` is supposed to remove it so you wouldn't still have it in readonly). So your code should not be needed. Instead, we need to figure out why `_activateContenteditable` doesn't apply `contenteditable="true"`. When it calls `getContentEditableAreas`, does it indeed receive the banner's div as part of the response? I see the PR has a big comment above `getContentEditableAreas`:
```js
// TODO other places in this file call this.options.getContentEditableAreas
// without the extension here. It does not seem to be a problem (it
// was like that before o_editor_banner elements were considered
// here), but we might want to review that.
```
It seems that this would be the most obvious cause. If `_activateContenteditable` doesn't receive the banner's div as part of the response from `getContentEditableAreas`, then `getContentEditableAreas` is called without the needed extension, and that's what needs fixing.

dhba-odoo replied:
In a typical scenario, `_activateContenteditable` is called on `_onMouseDown` or `historyRevertCurrentStep`. However, when creating a new banner, neither of these methods is invoked. Consequently, `_activateContenteditable` is not called, and the banner remains non-editable upon creation. To address this, I have call `_activateContenteditable` after the banner is inserted to ensure it make the banner editable.

Zynton commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -85,6 +85,20 @@ QUnit.module(
+                            <h1 placeholder="Heading 1" class=\"oe-hint\"><br></h1>
+                        </div>
+                    </div><p><br></p>`,
+                "Should Convert Paragraph Tag into Heading Tag Inside Newly Created Banner",
```
Same comment as before about the capitals.

Zynton commented on addons/web_editor/static/src/js/editor/odoo-editor/src/commands/deleteBackward.js:
```diff
@@ -101,7 +101,7 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false,
             parentEl.parentElement &&
             !parentEl.parentElement.isContentEditable &&
-            paragraphRelatedElements.includes(this.tagName) &&
+            this.tagName === 'P' &&
```
I wrote a comment on the task about this, I'm not sure about this spec. But if we keep it, this does seem like the right way to go about it.

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -1963,7 +1964,16 @@ export class Wysiwyg extends Component {
+                    range.selectNodeContents(container);
+                } else {
+                    let start = container.firstChild;
+                    while((!isVisible(start)) && start.nextSibling) {
```
```suggestion
                    while (start.nextSibling && !isVisible(start)) {
```

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -1963,7 +1964,16 @@ export class Wysiwyg extends Component {
+                    while((!isVisible(start)) && start.nextSibling) {
+                        start = start.nextSibling;
+                    }
+                    range.setStart(start, 0);
+                    range.setEnd(...endPos(container.lastChild));
```
What you needed to do to find the start node, how come you don't need to do it to find the end node as well?

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -1968,7 +1968,7 @@ export class Wysiwyg extends Component {
                 } else {
                     let start = container.firstChild;
-                    while((!isVisible(start)) && start.nextSibling) {
+                    while((!isVisible(start) || !start.isContentEditable) && start.nextSibling) {
```
```suggestion
                    while (start.nextSibling && !(isVisible(start) && start.isContentEditable)) {
```

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -1963,7 +1964,16 @@ export class Wysiwyg extends Component {
             if (container) {
                 const range = document.createRange();
-                range.selectNodeContents(container);
+                if (!container.hasChildNodes()) {
```
You make the fix at the shortcut level, but what if the user right clicked, then clicked "select all"?

dhba-odoo replied:
I overlooked this scenario; i will investigate and address it properly.

Zynton commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -1968,7 +1968,7 @@ export class Wysiwyg extends Component {
                     range.selectNodeContents(container);
                 } else {
                     let start = container.firstChild;
```
Same remark about right-click.

Zynton reviewed (changes requested):
Good work. A few remarks to address, detailed in my review.

dhba-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:
```diff
@@ -1867,6 +1867,12 @@ export function splitElement(element, offset) {
         index < offset ? before.appendChild(child) : after.appendChild(child);
         index++;
     }
+    // e.g.: <p>Test/banner</p> + ENTER <=> <p>Test</p><div class="o_editor_banner>...</div><p><br></p>
```
When a user creates a banner like in example, an issue arises where the after element appears empty without a line break `<br>`. Consequently, it becomes challenging to access the next line after the banner if needed.

Similarly, when creating a table using a similar approach, clicking at the end of the table 'type / for commands' powerbox hint will appear below the table. However, the cursor is still positioned at the end of the table, and any typing at that point inserts content outside the paragraph `(<p>)`. like, `<table></table>test<p></p>`
I have check the sanitizer and it's not adding `<br>` in those empty tags.

Zynton replied:
Ok indeed I see that `fillEmpty` (which does what you need here) is indeed called in various places where we need it, and only for one specific case in the sanitizer. Note to self: it would be good to refactor that.
In the meantime, it's ok to leave it here but you should use the util (`fillEmpty`) rather than duplicate the logic.

dhba-odoo replied:
I've applied the suggested changes.

dhba-odoo commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -0,0 +1,90 @@
+
+        QUnit.module("Banner");
+
+        QUnit.test("First element of o_editable is editable", async function (assert) {
```
All test cases in this file is related to banner, if its confusing i will change it to something meaningful

Zynton replied:
I mean it's not easy to tell from this title what it has to do with banner. I expect it to have a link with banner since it's in this file, but the title of this test is puzzling.

Zynton replied:
So, is the test more like "Insert a banner"?

dhba-odoo replied:
Yes, it's to check whether the new banner is properly inserted or not.

dhba-odoo commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -0,0 +1,90 @@
+                            <p placeholder=\"Type &quot;/&quot; for commands\" class=\"oe-hint oe-command-temporary-hint\"><br></p>
+                        </div>
+                    </div><p><br></p>`,
+                "Should Insert a Banner with Paragraph Tag",
```
I apologize for any confusion. Yes, it is followed by a paragraph.

Zynton replied:
OK then please rephrase :-)

Zynton commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -94,6 +94,34 @@ QUnit.module(
+                            <p><br></p>
+                        </div>
+                    </div><p placeholder=\"Type &quot;/&quot; for commands\" class=\"oe-hint oe-command-temporary-hint\"><br></p>`,
+                "should not remove banner when ctrl+a and backspace are preformed",
```
typo: "performed"

Zynton replied:
The typo is still there.

dhba-odoo replied:
done!

Zynton commented on addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:
```diff
@@ -830,15 +830,15 @@ export function getDeepestPosition(node, offset) {
             direction = offset < node.childNodes.length;
             next = node.childNodes[direction ? offset : offset - 1];
-        } else if (direction && next.nextSibling && !isBlock(next.nextSibling)) {
+        } else if (direction && next.nextSibling && (!isBlock(next.nextSibling) || !next.isContentEditable)) {
```
Semantics and style so feel free not to listen to me, but I prefer `!(isBlock(next.nextSibling) && next.isContentEditable)`. Most of the time I prefer `!(a && b)` over `!a || !b` because it's more intuitive: what we want is to not have the situation where these two things are true at the same time. It's harder to say it the other way. But that's me.

dhba-odoo replied:
I have removed these specific changes, so there is no longer a need to modify it.

Zynton commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -0,0 +1,90 @@
+            });
+            await makeView({
+                type: "form",
+                serverData,
```
By "here" I meant directly in `beforeEach`. I don't see a reason to keep it global.

dhba-odoo replied:
I have implemented the suggested changes!

Zynton commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -61,8 +61,37 @@ QUnit.module(
     function () {
 
-        QUnit.module("banner insertion tests");
+        QUnit.module("banner selection and backspace related tests");
```
No need to include "tests" in the name.

dhba-odoo replied:
done!

Zynton commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -61,8 +61,37 @@ QUnit.module(
+            );
+        });
 
+        QUnit.module("banner insertion tests");
```
Same.

dhba-odoo replied:
done!

Zynton reviewed (changes requested):
@dhba-odoo I saw you force pushed a few times two weeks ago but it seems you haven't yet addressed the review comments so I stopped re-reviewing. Are you planning on picking this up again soon?

dhba-odoo commented:
> @dhba-odoo I saw you force pushed a few times two weeks ago but it seems you haven't yet addressed the review comments so I stopped re-reviewing. Are you planning on picking this up again soon?

Yes, I'm currently working on it. After making changes to the code, I moved it for functional testing to identify any cases that may not be working as intended. Additionally, I've updated the entire test case.

shsa-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2203,7 +2203,8 @@ export class Wysiwyg extends Component {
                 `).childNodes[0];
                 this.odooEditor.execCommand('insert', bannerElement);
-                setSelection(bannerElement.querySelector('.o_editor_banner > div'), 0);
+                this.odooEditor._activateContenteditable();
+                setSelection(bannerElement.querySelector('.o_editor_banner > div > p'), 0);
```
sorry for the last minute input, i observed something which should be addressed as well. @dhba-odoo accessing the private method is not recommended, i have a suggestion which can be a workaround 
 1) by using 't-ref' to banner element and setting focus to it's <p>. 
2) or we can make '_activateContenteditable' a public method keeping it in utils. such that `wysiwyg` and `odooeditor` both can use it according to its convenience.
cc @Zynton what is your input on this ?

Zynton replied:
It's indeed not ideal and I'm a little confused as to why it's being accessed anyway if the goal is just to set the focus. Wouldn't `bannerElement.focus()` suffice? If not, how come?

dhba-odoo replied:
Here, `bannerElement.focus()` will not work because the `contentEditable` attribute for the`bannerElement` is `false`.

Zynton replied:
Ah I see that `contenteditable` is only set to true on the `div` that you want to edit (`.o_editor_banner > div`) because it's explicitely added as a content editable area:
```js
getContentEditableAreas: (...args) => {
    const areaEls = _getContentEditableAreas?.(...args) || [];
    const bannerEls = this.$editable[0].querySelectorAll('.o_editor_banner > div');
    return [...areaEls, ...bannerEls];
},
```
So indeed that only gets applied when `_activateContenteditable` is called, which means we need to make sure that gets called first... You could trigger a `mousedown` event for that but that's unclear and you'd have to deal with asynchronous stuff, so I'd argue in favor of making `_activateContenteditable` public. Although we can't change its name in stable so my suggestion is to add a method:
```js
activateContenteditable {
    this._activateContenteditable();
}
```
with a nice comment to justify it.

@sebgeelen is that ok for you?

sebgeelen replied:
sure

dhba-odoo replied:
Done!

Zynton commented on addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:
```diff
@@ -1868,6 +1868,11 @@ export function splitElement(element, offset) {
     }
+    // e.g.: <p>Test/banner</p> + ENTER <=> <p>Test</p><div class="o_editor_banner>...</div><p><br></p>
+    const blockEl = closestBlock(after);
+    if (isShrunkBlock(blockEl)) {
```
`fillEmpty` tests this too. Do you need to do it here as well so as not to insert zws? Otherwise you can remove this `if`.

dhba-odoo replied:
when I initially added the `fillEmpty` function, some test cases failed during local testing due to `zws` but now its working correctly. Therefore, it can be removed.

Zynton replied:
I checked and there is no reason we should ever enter the `if` that adds `zws` in this case since we're passing it a block and if it's shrunk it will add a `<br>`, making it visible. If it's not shrunk, there's no reason it would not be visible. So this change is good, thanks.

Zynton commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -2011,6 +2011,15 @@ export class OdooEditor extends EventTarget {
         return didDeselectTable;
     }
 
+    /**
+     * `activateContenteditable` serves as an interface for external use,
+     * allowing users to conveniently trigger `_activateContenteditable`
+     * from outside the odooEditor.
+     */
+    activateContenteditable() {
+        this._activateContenteditable();
+    }
+
```
Note: This should be adapted in master so we just rename the private function instead of duplicating.

Zynton replied:
Done here: https://github.com/odoo/odoo/pull/152825

Zynton commented on addons/web_editor/static/tests/banner_tests.js:
```diff
@@ -0,0 +1,85 @@
+    function () {
+
+        QUnit.module("insert a new banner");
+        QUnit.test("should insert a banner with paragraph tag", async function (assert) {
```
Didn't you say you meant "followed by a paragraph" rather than "with paragraph"? The distinction matters because "with paragraph" could mean that what matters in your test is that the banner _contains_ a paragraph (which it does), but from what I understand you want to make sure the banner is inserted and _also_ that a paragraph is inserted _after it_.

dhba-odoo replied:
I thought I changed it, but I probably accidentally reverted it while making other changes. Sorry for the inconvenience.

Zynton reviewed (changes requested):
Not an easy series of fixes but you're almost there!

dhba-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:
```diff
@@ -1868,6 +1868,9 @@ export function splitElement(element, offset) {
     }
+    // e.g.: <p>Test/banner</p> + ENTER <=> <p>Test</p><div class="o_editor_banner>...</div><p><br></p>
+    const blockEl = closestBlock(after);
+    blockEl && fillEmpty(blockEl);
```
done!

Zynton commented:
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
